pkg/instance/execution.go

```diff
@@ -0,0 +1,437 @@
+		if !running {
+			break
+		}
+		if err != nil {
+			return ErrCheckingIfInstanceStopped.WithParams(e.instance.k8sName).Wrap(err)
+		}
+
+		select {
+		case <-ctx.Done():
+			return ErrWaitingForInstanceTimeout.
+				WithParams(e.instance.k8sName).Wrap(ctx.Err())
+		case <-time.After(waitForInstanceRetry):
+			continue
+		}
+	}
+
+	return nil
+}
+
+// Stop stops the instance
+// CAUTION: In order to keep data of the instance, you need to use AddVolume() before.
+// This function can only be called in the state 'Started'
+func (e *execution) Stop(ctx context.Context) error {
+	if !e.instance.IsInState(StateStarted) {
+		return ErrStoppingNotAllowed.WithParams(e.instance.state.String())
+
+	}
+
+	if err := e.destroyPod(ctx); err != nil {
+		return ErrDestroyingPod.WithParams(e.instance.k8sName).Wrap(err)
+	}
+	e.instance.state = StateStopped
+	e.instance.sidecars.setStateForSidecars(StateStopped)
+	e.instance.Logger.Debugf("Set state of instance '%s' to '%s'", e.instance.k8sName, e.instance.state.String())
+
+	return nil
+}
+
+// Labels returns the labels for the instance
+func (e *execution) Labels() map[string]string {
+	return map[string]string{
+		labelAppKey:         e.instance.k8sName,
+		labelManagedByKey:   labelKnuuValue,
+		labelScopeKey:       e.instance.Scope,
+		labelTestStartedKey: e.instance.StartTime,
+		labelNameKey:        e.instance.name,
+		labelK8sNameKey:     e.instance.k8sName,
+		labelTypeKey:        e.instance.instanceType.String(),
+	}
+}
+
+// Destroy destroys the instance
+// This function can only be called in the state 'Started' or 'Destroyed'
+func (e *execution) Destroy(ctx context.Context) error {
+	if e.instance.state == StateDestroyed {
+		return nil
+	}
+
+	if !e.instance.IsInState(StateStarted, StateStopped) {
+		return ErrDestroyingNotAllowed.WithParams(e.instance.state.String())
+	}
+
+	if err := e.destroyPod(ctx); err != nil {
+		return ErrDestroyingPod.WithParams(e.instance.k8sName).Wrap(err)
+	}
+	if err := e.instance.resources.destroyResources(ctx); err != nil {
+		return ErrDestroyingResourcesForInstance.WithParams(e.instance.k8sName).Wrap(err)
+	}
+
+	err := e.instance.sidecars.applyFunctionToSidecars(
+		func(sidecar SidecarManager) error {
+			e.instance.Logger.Debugf("Destroying sidecar resources from '%s'", sidecar.Instance().k8sName)
+			return sidecar.Instance().resources.destroyResources(ctx)
+		})
+	if err != nil {
+		return ErrDestroyingResourcesForSidecars.WithParams(e.instance.k8sName).Wrap(err)
+	}
+
+	e.instance.state = StateDestroyed
+	e.instance.sidecars.setStateForSidecars(StateDestroyed)
+	e.instance.Logger.Debugf("Set state of instance '%s' to '%s'", e.instance.k8sName, e.instance.state.String())
+
+	return nil
+}
+
+func (e *execution) UpgradeImage(ctx context.Context, image string) error {
+	if !e.instance.IsInState(StateStarted) {
+		return ErrUpgradingImageNotAllowed.WithParams(e.instance.state.String())
+	}
+
+	return e.instance.build.setImageWithGracePeriod(ctx, image, nil)
+}
```

Comment:
Allow the dev to specify the grace period

Reply:
done